### ISSUE TITLE
mingens for PIDs

### ISSUE THREAD

**Issue by mahrud** [Macaulay2/M2#837]
See the code below:
```
restart
R = ZZ[i]/(i^2+1)
I = ideal"5,i+2"
mingens I
-- returns | 5 i+2 |
minimalPresentation I
-- returns ideal (5, 5)
```
This shouldn't happen since the ring of Gaussian integers is a PID, but I'm assuming it's happening because `i+2` is not homogeneous.

**Comment by DanGrayson** [comment]
I see no reason that shouldn't happen.  Try this instead:

```
Macaulay2, version 1.12
with packages: ConwayPolynomials, Elimination, IntegralClosure, InverseSystems,
               LLLBases, PrimaryDecomposition, ReesAlgebra, TangentCone

i1 : R = ZZ[i]/(i^2+1)

o1 = R

o1 : QuotientRing

i2 : I = ideal"5,i+2"

o2 = ideal (5, i + 2)

o2 : Ideal of R

i3 : smithNormalForm gens I

o3 = (| -5i+10 5   0 0 |, | -i+2 |, {0} | 1 0 -i-2 1   |)
      | 5      i+2 0 0 |  | 1    |  {1} | 0 1 5    i-2 |

o3 : Sequence
```

**Comment by mahrud** [comment]
Ahh, sorry, I wrote `mingens` in the title but used `gens gb` in the example. Also `minimalPresentation` gives `ideal (5, 5)`. Is that not unexpected?

**Comment by DanGrayson** [comment]
I think minimality computations succeed only in homogeneous situations.  @mikestillman ?

**Comment by mahrud** [comment]
Reviving this issue, which is partially related to what happens in #1660.

Note that I can make the ideal be homogeneous, and the output is still unexpected:
```m2
i1 : R = ZZ[i, DegreeRank => 0]/(i^2+1)

o1 = R

o1 : QuotientRing

i2 : I = ideal"5,i+2"

o2 = ideal (5, i + 2)

o2 : Ideal of R

i3 : isHomogeneous I

o3 = true

i4 : mingens I

o4 = | 5 i+2 |

             1       2
o4 : Matrix R  <--- R

i5 : trim I

o5 = ideal (5, i + 2)

o5 : Ideal of R

i6 : minimalPresentation I

o6 = ideal (5, 5)

o6 : Ideal of ZZ[]
```

**Comment by mahrud** [comment]
cc: @jchen419

**Comment by DanGrayson** [comment]
I guess you want the output on line o6 above to be `ideal 5`.

**Comment by jchen419** [comment]
@DanGrayson: shouldn't the output of o6 be `ideal(i+2)` instead? (as `5 = (2+i)*(2-i)`)

**Comment by DanGrayson** [comment]
No, `ideal 5` is an ideal in the ring `ZZ[]`, which has no variable called i in it.  (It's the quotient ring we're getting a minimal presentation of, getting a new ideal in a new ring with a quotient ring that's isomorphic to the original.)

**Comment by jchen419** [comment]
Ah, OK - then yes, it does seem like `ideal 5` should be returned instead.

**Comment by d-torrance** [comment]
> **Written by Claude** (Claude Opus 5, via Claude Code), not by @d-torrance, whose account posted it -- please weigh it accordingly.

Both forms still reproduce on 1.26.06-40-gd8e86d689d, and the thread turns out to have missed
that M2 already contains the strategy @DanGrayson reached for in his first reply.

### Still there, and homogeneity is not the obstacle

```m2
i1 : R = ZZ[i]/(i^2+1); I = ideal"5,i+2";

i3 : minimalPresentation I

o3 = ideal (5, 5)

o3 : Ideal of ZZ[]
```

and with @mahrud's homogeneous variant, `ZZ[i, DegreeRank => 0]/(i^2+1)`, `isHomogeneous I` is
`true` and the answer is still `ideal (5, 5)`. So the 2018 guess that "minimality computations
succeed only in homogeneous situations" is not what is happening here.

Worth being exact about the severity: `ideal (5, 5)` and `ideal 5` are the same ideal, so nothing
is *wrong* — the result is redundant, not incorrect. `mingens` and `trim` both return
`5, i+2` unchanged, which is also a correct generating set and also not minimal.

### There is a PID strategy; this ring does not reach it

`M2/Macaulay2/m2/matrix2.m2:301` registers exactly the `smithNormalForm` approach from the reply
above:

```m2
addHook((trim, Module), Strategy => "PID",
    (opts, M) -> (
	R := ring M;
	if instance(R,PolynomialRing) and numgens R === 1 and isField coefficientRing R and not isHomogeneous M then trimPID M
	)
    )
```

`ZZ[i]/(i^2+1)` fails that guard twice over. It is a `QuotientRing` rather than a
`PolynomialRing`, and its coefficient ring is `ZZ` rather than a field. So the operative notion
of "PID" here is "univariate polynomial ring over a field" — narrower than the mathematical one,
and narrower than the title of this issue assumes. The Gaussian integers are a PID; M2 just has
no route to treating them as one.

That seems the most useful thing to record: the gap is a guard, not a missing algorithm, and
`trimPID` immediately above it is already written in terms of `smithNormalForm` and a change of
basis.

### Not the same as #44

#44, "minimalPresentation for modules over ZZ", is the same theme and a different mechanism —
there the ring is `ZZ[x,y]`, which is not a PID, and @mikestillman's assessment is that it "might
require a rewrite of some of the GB over ZZ code". Widening a guard would not touch it. Noting
the pair here rather than merging them.

<sub>Triaged in a sweep of the open issues; source of truth: [`bug-triage/issues.tsv`](https://github.com/d-torrance/M2/blob/bug-triage/bug-triage/issues.tsv).</sub>
